File: binary_tree_right_side_view.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
    res=[]
    def printTree(self, root):
        if not root:
            return []
        if root:
            self.res.append(root.val)
            if root.left:
                self.printTree(root.left)
            if root.right:
                self.printTree(root.right)
        return self.res

class Solution:
    def rightSideView(self, root):
        
        def dfs(root,depth):
            if root:
                if self.cur < depth:
                    self.cur = depth
                    res.append(root.val)

                dfs(root.right,depth+1)
                dfs(root.left,depth+1)
        res = []
        if root:
            self.cur = -1
            dfs(root,0)
        return res
if __name__ == '__main__':
    sol = Solution()
    tree = TreeNode(None)
    root= TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.right = TreeNode(4)
    #root.right.right = TreeNode(5)
    print sol.rightSideView(root)
    

File: Algorithm/CPP/binary_tree_right_side_view.cpp
# include <stdio.h>
# include <iostream>
# include <vector>
# include <string>
# include <unordered_map>
# include <cctype>
# include <stack>
# include <algorithm>
# include <time.h>
# include <queue>
using namespace std;

struct TreeNode {
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
	vector<int> rightSideView(TreeNode* root) 
	{
		vector<int> res;
		dfs(res, root, 1);
		return res;

	}
	void dfs(vector<int> & res, TreeNode* root, int depth)
	{
		if (root)
		{
			if (res.size() < depth) // the new level, should build a new vector<int> to the res 
				res.push_back(root->val);
			dfs(res, root->right, depth + 1);
			dfs(res, root->left, depth + 1);
			
		}
	}

};
int  main()
{
	Solution sol;
	TreeNode *root1;
	root1 = new TreeNode(4);
	root1->left = new  TreeNode(2);
	root1->right = new TreeNode(5);
	root1->left->left = new TreeNode(1);
	root1->left->right = new TreeNode(3);
	for (auto x : sol.rightSideView(root1))
	{

		cout << x << " ";

	}
	cout << endl;
	system("pause");
	return 0;
}
